### ISSUE TITLE
How to fine tune roberta for multi-label classification?

### ISSUE THREAD

**Issue by neelansh** [pytorch/fairseq#2169]
Sorry new to fairseq. I would like to know how can we fine-tune roberta model for multilabel classification. Which fairseq dataset would be best for the task? do I need a custom task or can I use sentence prediction task?

**Comment by villmow** [comment]
https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.custom_classification.md

**Comment by bdropuljic** [comment]
@villmow the readme you linked is only for solution in which you can have only one label per sample (the multiclass problem). But what about the situation in which you can have multiple labels per sample? E.g. you have a sentence that is being labeled with 2 or 3 categories?

**Comment by villmow** [comment]
Yes you're right. Sorry I missed that. I think then you'll need a custom task. The sentence_prediction task simply uses a CrossEntropyLoss, which is probably not what you want for multi-label classification. You'd probably want a BinaryCrossEntropy loss...

**Comment by neelansh** [comment]
Yes, I also ended up creating a custom task and criterion for using Binarycrossentropy. But finding the right dataset class for loading the multilabel data was tricky, I ended up using the LanguagePairDataset and one hot encoded the labels to a fixed length.

**Comment by pstjohn** [comment]
@neelansh, any chance you might be willing to share a gist for how you loaded the data?

**Comment by neelansh** [comment]
@pstjohn sure. Let me dig it out. and would post it soon.

**Comment by neelansh** [comment]
@pstjohn here is the code for loading the multilabel data. You need to create a custom task where you can define this data loader function and a custom criterion that uses binary cross entropy loss. you can register both these classes using `@register_task` and `@register_criterion`  decorators. 

The following is the load_data set definition for the custom task
`def load_dataset(self, split, combine=False, **kwargs):
        """Load a given dataset split (e.g., train, valid, test)."""
        split_path = os.path.join(self.args.data, 'input0', split)
        input0 = data_utils.load_indexed_dataset(
                split_path,
                self.source_dictionary,
                self.args.dataset_impl,
                combine=combine,
            )
        
        if self.args.init_token is not None:
            input0 = PrependTokenDataset(input0, self.args.init_token)
            
        src_tokens = input0
        
        with data_utils.numpy_seed(self.args.seed):
            shuffle = np.random.permutation(len(src_tokens))

        if self.args.truncate_sequence:
            src_tokens = TruncateDataset(src_tokens, self.args.max_positions)
        
        src_dataset = src_tokens
        
        labels, label_lengths = [], []
        with open(os.path.join(self.args.data, 'label', split+".label")) as file:
            for line in file:
                line = line.strip()
                line = line.split(" ")
                label = [self.label_vocab.add_symbol(l) for l in line]
                
                if(len(label) < self.args.max_target_positions):
                    label = label + [self.label_vocab.add_symbol('<pad>')]*(self.args.max_target_positions-len(label))
                
                label = label[:self.args.max_target_positions]
                
                label = torch.tensor(label)
                labels.append(label)
                label_lengths.append(len(label))
        
        assert len(src_dataset) == len(labels)
        print('| {} {} {} examples'.format(self.args.data, split, len(src_dataset)))
        self.datasets[split] = LanguagePairDataset(
            src=src_dataset,
            src_sizes=src_dataset.sizes,
            src_dict=self.input_vocab,
            tgt=labels,
            tgt_sizes=torch.tensor(label_lengths),
            tgt_dict=self.label_vocab,
            left_pad_source=False,
            max_source_positions=self.args.max_positions,
            max_target_positions=self.args.max_target_positions,
            input_feeding=False,
        )`

following is the code snippet for using binary_cross_entropy_with_logits loss for multilabel classification. This needs to be added to the forward definition in your custom criterion class

`
targets = F.one_hot(targets.long(), num_classes=logits.size()[-1]+4)
targets = targets.sum(dim=1)
targets = targets[:, 4:]
loss = F.binary_cross_entropy_with_logits(logits, targets.float(), reduction='sum')
`

**Comment by sipan17** [comment]
Hi @neelansh, in which file(s) did you add/modify the code?
